### PR TITLE
PR: Fix issue where replace all did not respect whole words filter

### LINKED ISSUE #20403: Find and replace "Replace all occurences" ignores "Only search for whole words"
### Issue Report Checklist

* [x] Searched the [issues page](https://github.com/spyder-ide/spyder/issues?q=is%3Aissue) for similar reports
* [x] Read the relevant sections of the [Spyder Troubleshooting Guide](https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ) and followed its advice
* [x] Reproduced the issue after updating with ``conda update spyder`` (or ``pip``, if not using Anaconda)
* [ ] Could not reproduce inside ``jupyter qtconsole`` (if console-related)
* [ ] Tried basic troubleshooting (if a bug/error)
    * [x] Restarted Spyder
    * [ ] Reset preferences with ``spyder --reset``
    * [ ] Reinstalled the latest version of [Anaconda](https://www.anaconda.com/download/)
    * [ ] Tried the other applicable steps from the Troubleshooting Guide
* [ ] Completed the **Problem Description**, **Steps to Reproduce** and **Version** sections below

## Problem Description
find and replace ignores "Only search for whole words" when doing "Replace all occurences". Only the correct occurences are highlighted and if I replace them one by one then all is fine. The problem is just with "Replace all occurences".

### What steps reproduce the problem?

1. Write in a new file
```
aaa
aaa2
```
2. Open find and replace. Write `aaa` in the find line. Write `aaa3` in the replace line.
3. Check "Only search for whole words". Only `aaa` is highlighted.
4. Press "Replace all occurences".

### What is the expected output? What do you see instead?
Expected:
```
aaa3
aaa2
```
I get
```
aaa3
aaa32
```

## Versions

* Spyder version: 5.4.1 (although I'm quite sure it was 5.3.x before updating as requested in the issue guidelines)
* Python version: 3.10.8 64-bit
* Qt version: 5.15.6
* PyQt version: 5.15.7
* Operating System name/version: Linux 5.4.0-137-generic 

### Dependencies

```
# Mandatory:
atomicwrites >=1.2.0          :  1.4.1 (OK)
chardet >=2.0.0               :  5.1.0 (OK)
cloudpickle >=0.5.0           :  2.2.0 (OK)
cookiecutter >=1.6.0          :  2.1.1 (OK)
diff_match_patch >=20181111   :  20200713 (OK)
intervaltree >=3.0.2          :  3.0.2 (OK)
IPython >=7.31.1;<9.0.0       :  8.8.0 (OK)
jedi >=0.17.2;<0.19.0         :  0.18.2 (OK)
jellyfish >=0.7               :  0.9.0 (OK)
jsonschema >=3.2.0            :  4.17.3 (OK)
keyring >=17.0.0              :  23.13.1 (OK)
nbconvert >=4.0               :  7.2.8 (OK)
numpydoc >=0.6.0              :  1.5.0 (OK)
parso >=0.7.0;<0.9.0          :  0.8.3 (OK)
pexpect >=4.4.0               :  4.8.0 (OK)
pickleshare >=0.4             :  0.7.5 (OK)
psutil >=5.3                  :  5.9.4 (OK)
pygments >=2.0                :  2.14.0 (OK)
pylint >=2.5.0;<3.0           :  2.15.10 (OK)
pylint_venv >=2.1.1           :  2.3.0 (OK)
pyls_spyder >=0.4.0           :  0.4.0 (OK)
pylsp >=1.7.0;<1.8.0          :  1.7.1 (OK)
pylsp_black >=1.2.0           :  1.2.1 (OK)
qdarkstyle >=3.0.2;<3.1.0     :  3.0.3 (OK)
qstylizer >=0.2.2             :  0.2.2 (OK)
qtawesome >=1.2.1             :  1.2.2 (OK)
qtconsole >=5.4.0;<5.5.0      :  5.4.0 (OK)
qtpy >=2.1.0                  :  2.3.0 (OK)
rtree >=0.9.7                 :  1.0.1 (OK)
setuptools >=49.6.0           :  66.0.0 (OK)
sphinx >=0.6.6                :  6.1.3 (OK)
spyder_kernels >=2.4.1;<2.5.0 :  2.4.2 (OK)
textdistance >=4.2.0          :  4.5.0 (OK)
three_merge >=0.1.1           :  0.1.1 (OK)
watchdog >=0.10.3             :  2.2.1 (OK)
xdg >=0.26                    :  0.28 (OK)
zmq >=22.1.0                  :  25.0.0 (OK)

# Optional:
cython >=0.21                 :  None (NOK)
matplotlib >=3.0.0            :  3.6.2 (OK)
numpy >=1.7                   :  1.24.1 (OK)
pandas >=1.1.1                :  1.5.2 (OK)
scipy >=0.17.0                :  1.10.0 (OK)
sympy >=0.7.3                 :  None (NOK)```

### PR DESCRIPTION
**Pull request by mrclary**
## Description of Changes

* Update replace pattern to respect whole words filter
* Added regression test

### Issue(s) Resolved

Fixes #20403

### CODE REVIEW

File: spyder/widgets/findreplace.py [hunk 1]
```diff
@@ -666,24 +665,31 @@ def replace_find_all(self):
...
-                # Do nothing with an invalid regexp
-                return
+
+        re_pattern = None
+        if self.re_button.isChecked():
+            pattern = search_text
         else:
-            re_pattern = re.compile(re.escape(search_text), flags=re_flags)
+            pattern = re.escape(search_text)
+            replace_text = replace_text.replace('\\', r'\\')
```
Comment (impact27) [inline spyder/widgets/findreplace.py:680]: Why is that necessary?

Comment (impact27) [reply spyder/widgets/findreplace.py:680]: Maybe `replace_text = re.escape(replace_text)`?
In the same vein, is there already a test for regexp replace?

eg: 
`a123a`
search: `a(\d+)a`
replace: `b\1b`
result: `b123b`

Comment (impact27) [reply spyder/widgets/findreplace.py:680]: If a similar function doesn't exist for sub text, I would just add a comment to explain.
Maybe:
`replace_text.replace(r'\', r'\\')`
is more legible than
`replace_text.replace('\\', r'\\')`
?

Comment (mrclary) [reply spyder/widgets/findreplace.py:680]: > Why is that necessary?
> Maybe replace_text = re.escape(replace_text)?

I don't know for sure if it is necessary. I was just imitating the procedures used in `replace_find_selection`. As far as I can tell, `re.escape` does exactly the same thing and I would prefer that to `replace`. Perhaps I should make the correction in `replace_find_selection` as well?

> In the same vein, is there already a test for regexp replace?

No there is not. I will add your suggested test to `test_replace_all`.

Comment (mrclary) [reply spyder/widgets/findreplace.py]: Actually, it looks like it has to remain `replace_text.replace('\\', r'\\')`.
`re.escape` cannot be used since it escapes _all_ special characters, not just `\`.
Furthermore, `replace_text.replace(r'\', r'\\')` results in code analysis reporting "Unexpected character after line continuation character".

Comment (impact27) [reply spyder/widgets/findreplace.py:680]: Reading re.sub:

repl can be a string or a function; if it is a string, any backslash escapes in it are processed. That is, \n is converted to a single newline character, \r is converted to a carriage return, and so forth. Unknown escapes of ASCII letters are reserved for future use and treated as errors. Other unknown escapes such as \& are left alone. Backreferences, such as \6, are replaced with the substring matched by group 6 in the pattern.

It looks like you are correct. You could add a comment saying: escape backslashes as they are processed by re.sub

**Comment by ccordoba12** [comment]
@mrclary, I changed the base branch of your PR to `master` because we agreed that from now on only critical things or important bug fixes should be addressed for Spyder 5.

**Comment by mrclary** [comment]
> @mrclary, I changed the base branch of your PR to `master` because we agreed that from now on only critical things or important bug fixes should be addressed for Spyder 5. And I don't think this is really critical.

I'll rebase my branch off master as well.

**Review by ccordoba12 — COMMENTED** [review]
Thanks @mrclary for your work on this!

File: spyder/widgets/tests/test_findreplace.py [hunk 2]
```diff
@@ -100,6 +100,52 @@ def test_replace_selection(findreplace_editor, qtbot):
...
+    editor = findreplace_editor.editor
+    findreplace = findreplace_editor.findreplace
+    findreplace.show_replace()
+
+    findreplace.search_text.setCurrentText('a')
+    findreplace.replace_text.setCurrentText('x')
+
+    # Replace all
+    editor.set_text('a\naa')
+    expected = 'x\u2029xx'  # Note: \n is replaced with \u2029
```
[the remark this answers is not in the document]
Comment (mrclary) [reply spyder/widgets/tests/test_findreplace.py:116]: Oh, I see. `findreplace.replace_find` corrects for this, but `findreplace.replace_find_all` and `findreplace.replace_find_selection` do not. Perhaps `replace(u'\u2029', eol_char)` (`findreplace.py:640`) should be applied in these methods as well? Then the unit test would not need `\u2029`.

[the remark this answers is not in the document]
Comment (mrclary) [reply spyder/widgets/tests/test_findreplace.py:116]: Nevermind, I just need to use `editor.toPlainText()`

File: spyder/widgets/findreplace.py [hunk 3]
```diff
@@ -666,24 +665,32 @@ def replace_find_all(self):
...
-                return
+
+        if self.re_button.isChecked():
+            pattern = search_text
         else:
-            re_pattern = re.compile(re.escape(search_text), flags=re_flags)
+            pattern = re.escape(search_text)
+            replace_text = replace_text.replace('\\', r'\\')
+        if word:  # match whole words only
+            pattern = r'\b{pattern}\b'.format(pattern=pattern)
```
Comment (ccordoba12) [inline spyder/widgets/findreplace.py:681]:
```suggestion

        # Match whole words only
        if word:
            pattern = r'\b{pattern}\b'.format(pattern=pattern)
```

**Review by ccordoba12 — COMMENTED** [review]
One last stylistic change that I missed before, then this should be ready for me.

**Comment by ccordoba12** [comment]
@impact27, do you have something else to say about this one?

**Comment by impact27** [comment]
Looks good to me, I would just add a comment explaining replace_text = replace_text.replace('\\', r'\\') (re.sub processes backslashes so it must be escaped)

**Review by ccordoba12 — APPROVED** [review]
Thanks @mrclary!
